File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.scala
import scala.collection.mutable.ListBuffer
import scala.collection.mutable.Map

object Solution {
  def findSubstring(s: String, words: Array[String]): List[Int] = {
    val result = ListBuffer[Int]()

    if (s == null || words == null || words.isEmpty) {
      return result.toList
    }

    val n = s.length
    val m = words.length
    val len = words(0).length

    val map = Map[String, Int]()
    for (word <- words) {
      map(word) = map.getOrElse(word, 0) + 1
    }

    var i = 0
    while (i <= n - m * len) {
      val seen = Map[String, Int]()
      var j = 0
      var found = true

      while (j < m && found) {
        val k = i + j * len
        val sub = s.substring(k, k + len)

        if (!map.contains(sub)) {
          found = false
        } else {
          seen(sub) = seen.getOrElse(sub, 0) + 1

          if (seen(sub) > map(sub)) {
            found = false
          }
        }
        j += 1
      }

      if (found) {
        result += i
      }
      i += 1
    }
    result.toList
  }
}

File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.java
class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> result = new ArrayList<>();
        if (s == null || words == null || words.length == 0) {
            return result;
        }

        int n = s.length();
        int m = words.length;
        int len = words[0].length();

        Map<String, Integer> map = new HashMap<>();
        for (String word : words) {
            map.put(word, map.getOrDefault(word, 0) + 1);
        }

        for (int i = 0; i < n - m * len + 1; i++) {
            Map<String, Integer> seen = new HashMap<>();
            int j = 0;
            for (; j < m; j++) {
                int k = i + j * len;
                String sub = s.substring(k, k + len);
                if (!map.containsKey(sub)) {
                    break;
                }
                seen.put(sub, seen.getOrDefault(sub, 0) + 1);
                if (seen.get(sub) > map.get(sub)) {
                    break;
                }
            }
            if (j == m) {
                result.add(i);
            }
        }
        return result;
    }
}